src/hash.cr

```diff
@@ -118,6 +290,479 @@ class Hash(K, V)
+      if hash == entry.hash && entry.key == key
+        set_entry(index, Entry(K, V).new(entry.hash, entry.key, value))
+        return true
+      end
+    end
+
+    # If full, resize. Otherwise we have space so add as last.
+    if entries_full?
+      resize
+      false
+    else
+      add_entry_and_increment_size(hash, key, value)
+      true
+    end
+  end
+
+  # Implementation of deleting a key.
+  # Returns the deleted Entry, if it existed, `nil` otherwise.
+  private def delete_impl(key) : Entry(K, V)?
+    # Empty hash table, nothing to do
+    if @indices_size_pow2 == 0
+      return nil
+    end
+
+    hash = key_hash(key)
+
+    # No indices allocated yet so do linear scan
+    if @indices.null?
+      return delete_linear_scan(key, hash)
+    end
+
+    # Fit hash into `@indices` size
+    index = fit_in_indices(hash)
+    while true
+      entry_index = get_index(index)
+
+      # If we find an empty index slot, there's no such key
+      if entry_index == -1
+        return nil
+      end
+
+      # We found a non-empty slot, let's see if the key we have matches
+      entry = get_entry(entry_index)
+      if entry.matches?(hash, key)
+        delete_entry_and_update_counts(entry_index)
+        return entry
+      else
+        # If it doesn't, check the next index...
+        index = next_index(index)
+      end
+    end
+  end
+
+  # Delete by doing a linear scan over `@entries`.
+  # Returns the deleted Entry, if it existed, `nil` otherwise.
+  private def delete_linear_scan(key, hash) : Entry(K, V)?
+    each_entry_with_index do |entry, index|
+      if entry.matches?(hash, key)
+        delete_entry_and_update_counts(index)
+        return entry
+      end
+    end
+
+    nil
+  end
+
+  # Finds an entry with the given key.
+  protected def find_entry(key) : Entry(K, V)?
+    # Empty hash table so there's no way it's there
+    if @indices_size_pow2 == 0
+      return nil
+    end
+
+    # No indices allocated yet so do linear scan
+    if @indices.null?
+      return find_entry_linear_scan(key)
+    end
+
+    hash = key_hash(key)
+
+    # Fit hash into `@indices` size
+    index = fit_in_indices(hash)
+    while true
+      entry_index = get_index(index)
+
+      # If we find an empty index slot, there's no such key
+      if entry_index == -1
+        return nil
+      end
+
+      # We found a non-empty slot, let's see if the key we have matches
+      entry = get_entry(entry_index)
+      if entry.matches?(hash, key)
+        # It does!
+        return entry
+      else
+        # Nope, move on to the next slot
+        index = next_index(index)
+      end
+    end
+  end
+
+  # Finds an Entry with the given key by doing a linear scan.
+  private def find_entry_linear_scan(key) : Entry(K, V)?
+    # If we have less than 8 elements we avoid computing the hash
+    # code and directly compare the keys (might be cheaper than
+    # computing a hash code of a complex structure).
+    if entries_size <= 8
```

Comment:
I doubt it is useful. I could be mistaken.
Was it benchmarked?

Reply:
Yes. I just benchmarked it again because I introduced that optimization some time ago, before introducing more optimizations and changing other things.

I ran the same bencharmks as in my main post but only for String. The benchmark for `insert` don't have a difference. For `read` it gives this (old is without the optimization, new is with it):

```
old, read (type: String, size: 1)  70.65M ( 14.15ns) (± 2.49%)  0.0B/op   2.68× slower
new, read (type: String, size: 1) 189.05M (  5.29ns) (± 3.89%)  0.0B/op        fastest
old, read (type: String, size: 2)  39.07M ( 25.59ns) (± 2.05%)  0.0B/op   2.26× slower
new, read (type: String, size: 2)  88.16M ( 11.34ns) (± 4.01%)  0.0B/op        fastest
old, read (type: String, size: 3)  26.56M ( 37.65ns) (± 1.84%)  0.0B/op   2.00× slower
new, read (type: String, size: 3)  53.24M ( 18.78ns) (± 1.89%)  0.0B/op        fastest
old, read (type: String, size: 4)  19.98M ( 50.06ns) (± 2.17%)  0.0B/op   1.72× slower
new, read (type: String, size: 4)  34.35M ( 29.11ns) (± 3.70%)  0.0B/op        fastest
old, read (type: String, size: 5)  16.00M ( 62.49ns) (± 1.61%)  0.0B/op   1.49× slower
new, read (type: String, size: 5)  23.78M ( 42.05ns) (± 2.05%)  0.0B/op        fastest
old, read (type: String, size: 6)  13.11M ( 76.25ns) (± 1.49%)  0.0B/op   1.39× slower
new, read (type: String, size: 6)  18.17M ( 55.02ns) (± 2.68%)  0.0B/op        fastest
old, read (type: String, size: 7)  11.21M ( 89.24ns) (± 1.37%)  0.0B/op   1.22× slower
new, read (type: String, size: 7)  13.71M ( 72.93ns) (± 3.94%)  0.0B/op        fastest
old, read (type: String, size: 8)   9.57M (104.47ns) (± 3.21%)  0.0B/op   1.16× slower
new, read (type: String, size: 8)  11.14M ( 89.77ns) (± 3.77%)  0.0B/op        fastest
```

So you can see it's a big optimization. You can also see it becomes faster and faster to start compare the hash first when there are more and more elements, starting from 9 it's almost the same.

I wonder if this optimization could also be applied in the Ruby code... 🤔 

I might give it a try later (in Ruby), now I'm curious!

Reply:
Actually, I think if the key type is more complex and `==` takes more time then it will be slower. So maybe it should also be applied for a few cases... maybe just do it if the key is a String because we know it's true.

Reply:
I just tried it with keys being `Array(String)` and the differences are actually much bigger, meaning that it's faster to avoid checking the `hash` code altogether for some reason... probably because to compute the hash code you need to consider the entire structure, but to compare elements you bail out as soon as you find a difference.